src/driver-cmstorm.c

```diff
@@ -0,0 +1,793 @@
+
+	do
+	{
+		ret = write_chunk(device, profile, offset, min(length, CMSTORM_BLOCK_SIZE), buf);
+		if (ret < 0)
+		{
+			return ret;
+		}
+		offset += CMSTORM_BLOCK_SIZE;
+		length -= CMSTORM_BLOCK_SIZE;
+		buf += CMSTORM_BLOCK_SIZE;
+	}
+	while (ret == 0 && length > 0);
+
+	return 0;
+}
+
+
+int read_profile(struct ratbag_device *device, int profile, struct cmstorm_profile *out)
+{
+	int ret = read_chunks(device, profile, 0, CMSTORM_PROFILE_SIZE, (uint8_t*) out);
+	if (ret < 0)
+		return ret;
+
+	//fixup the 16-bit integers
+	out->macro_length = be16toh(out->macro_length);
+	for (int i = 0; i < 9; i++)
+	{
+		out->buttons[i].function = be16toh(out->buttons[i].function);
+	}
+
+	//load the macros if there are any
+	if (out->macro_length > 0) {
+		out->macros = (struct cmstorm_macro*) zalloc(out->macro_length);
+
+		read_chunks(device, profile, 0x100, out->macro_length, (uint8_t*) out->macros);
+
+		for (int i = 0; i < out->macro_length / sizeof(struct cmstorm_macro); i++) {
+			out->macros[i].delay = be16toh(out->macros[i].delay);
+		}
+	}
+	return 0;
+}
+
+int write_profile(struct ratbag_device *device, int profile, struct cmstorm_profile *out)
+{
+	//fixup the 16-bit integers
+	int remaining_length = out->macro_length;
+	out->macro_length = htobe16(out->macro_length);
+	for (int i = 0; i < 9; i++) {
+		out->buttons[i].function = htobe16(out->buttons[i].function);
+	}
+
+	if (remaining_length > 0) {
+		for (int i = 0; i < out->macro_length / sizeof(struct cmstorm_macro); i++) {
+			out->macros[i].delay = htobe16(out->macros[i].delay);
+		}
+	}
+
+	int ret = write_chunks(device, profile, 0, CMSTORM_PROFILE_SIZE, (uint8_t*) out);
+	if (ret < 0)
+		return ret;
+
+	//load the macros if there are any
+	if (remaining_length > 0) {
+		write_chunks(device, profile, 0x100, remaining_length, (uint8_t*) out->macros);
+	}
+	return 0;
+}
+
+static int
+cmstorm_test_hidraw(struct ratbag_device *device)
+{
+	return ratbag_hidraw_has_report(device, CMSTORM_REPORT_ID);
+}
+
+static int
+cmstorm_probe(struct ratbag_device *device)
+{
+	struct ratbag_profile *profile = NULL;
+	struct ratbag_resolution *resolution;
+	struct ratbag_button *button;
+	struct ratbag_led *led;
+	int rc;
+	_cleanup_(dpi_list_freep) struct dpi_list *dpilist = NULL;
+	_cleanup_(freep) struct dpi_range *dpirange = NULL;
+	int selected_profile;
+
+	unsigned int report_rates[] = { 125, 250, 500, 1000 };
+
+	rc = ratbag_find_hidraw(device, cmstorm_test_hidraw);
+	if (rc)
+		return rc;
+
+	ratbag_device_init_profiles(device,
+					CMSTORM_NUM_PROFILES,
+					CMSTORM_NUM_DPI,
+					CMSTORM_NUM_BUTTONS,
+					CMSTORM_NUM_LEDS);
+
+	ratbag_device_set_capability(device, RATBAG_DEVICE_CAP_BUTTON);
+	ratbag_device_set_capability(device, RATBAG_DEVICE_CAP_BUTTON_KEY);
+	ratbag_device_set_capability(device, RATBAG_DEVICE_CAP_BUTTON_MACROS);
+
+	dpilist = ratbag_device_data_cmstorm_get_dpi_list(device->data);
+
+	get_selected_profile(device, &selected_profile);
+
+	struct cmstorm_profile cprofile;
```

Comment:
I always have a hard time with variable declarations in the middle of the code. I know it can be useful, but it hurts my eye every time.